tools/pytorch2onnx.py

```diff
@@ -1,124 +1,143 @@
+    one_img = torch.from_numpy(one_img).unsqueeze(0).float()
+    (_, C, H, W) = input_shape
+    one_meta = {
+        'img_shape': (H, W, C),
+        'ori_shape': (H, W, C),
+        'pad_shape': (H, W, C),
+        'filename': '<demo>.png',
+        'scale_factor': 1.0,
+        'flip': False
+    }
+    # onnx.export does not support kwargs
+    origin_forward = model.forward
+    model.forward = partial(
+        model.forward, img_metas=[[one_meta]], return_loss=False)
+    # pytorch has some bug in pytorch1.3, we have to fix it
+    # by replacing these existing op
+    register_extra_symbolics(opset_version)
+    torch.onnx.export(
+        model, ([one_img]),
+        output_file,
+        export_params=True,
+        keep_initializers_as_inputs=True,
+        verbose=show,
+        opset_version=opset_version)
+    model.forward = origin_forward
+    print(f'Successfully exported ONNX model: {output_file}')
+    if verify:
+        # check by onnx
+        onnx_model = onnx.load(output_file)
+        onnx.checker.check_model(onnx_model)
+
+        # check the numerical value
+        # get pytorch output
+        pytorch_result = model([one_img], [[one_meta]], return_loss=False)
+
+        # get onnx output
+        input_all = [node.name for node in onnx_model.graph.input]
+        input_initializer = [
+            node.name for node in onnx_model.graph.initializer
+        ]
+        net_feed_input = list(set(input_all) - set(input_initializer))
+        assert (len(net_feed_input) == 1)
+        sess = rt.InferenceSession(output_file)
+        from mmdet.core import bbox2result
+        det_bboxes, det_labels = sess.run(
+            None, {net_feed_input[0]: one_img.detach().numpy()})
+        # only compare a part of result
+        bbox_results = bbox2result(det_bboxes, det_labels, 1)
+        onnx_results = bbox_results[0]
+        assert np.allclose(
+            pytorch_result[0][:, 4], onnx_results[:, 4]
+        ), 'The outputs are different between Pytorch and ONNX'
+        print('The numerical values are same between Pytorch and ONNX')
 
 
 def parse_args():
-    parser = argparse.ArgumentParser(
-        description='MMDet pytorch model conversion to ONNX')
+    parser = argparse.ArgumentParser(description='Convert MMDet to ONNX')
     parser.add_argument('config', help='test config file path')
     parser.add_argument('checkpoint', help='checkpoint file')
+    parser.add_argument('--input_img', type=str, help='Images for input')
+    parser.add_argument('--show', action='store_true', help='show onnx graph')
+    parser.add_argument('--output_file', type=str, default='tmp.onnx')
+    parser.add_argument('--opset_version', type=int, default=11)
     parser.add_argument(
-        '--out', type=str, required=True, help='output ONNX filename')
+        '--verify',
+        action='store_true',
+        help='verify the onnx model output against pytorch output')
```

Comment:
Mean and std may be added as an argument.

Reply:
Done